File: Problems/programmers_프렌즈4블록.py
#https://school.programmers.co.kr/learn/courses/30/lessons/17679
#Implementation

def solution(m, n, board):
    answer = 0
    for i in range(len(board)):
        board[i] = list(board[i]) #string은 list 씌우면 문자 하나씩 떨어진다.
        
    while True:
        poplist = []
        for i in range(m-1):
            for j in range(n-1):
                if board[i][j] == 0:
                    continue
                elif board[i][j] == board[i+1][j] == board[i][j+1] == board[i+1][j+1] != 0:
                    poplist.extend([(i,j), (i+1,j), (i,j+1), (i+1,j+1)]) #list안의 tuple들을 각각 원소로 추가

        poplist = list(set(poplist)) # 중복체크
        if len(poplist) == 0: # 터질게 없으면 중단
            break
            
        answer += len(poplist)
        for x, y in poplist:
            board[x][y] = 0
    	
        for j in range(n):
            for i in range(m-1, -1, -1):
                if board[i][j] == 0:
                    x = i-1
                    while x>-1 and board[x][j]==0:
                        x -= 1

                    if x>=0:
                        board[i][j], board[x][j] = board[x][j], 0 #맞교환이 아니더라도 파이썬은 이렇게 쓸 수 있지!!
    return answer

File: Problems/cpp/programmers_프렌즈4블록.cpp
//https://school.programmers.co.kr/learn/courses/30/lessons/17679
//Implementation

#include <iostream>
#include <string>
#include <vector>
#include <set>

using namespace std;

int solution(int m, int n, vector<string> board) {
    int answer = 0;

    while (true){
        set<pair<int, int>> posSet;
        
        for (int i=0; i<m-1; i++){
            for (int j=0; j<n-1; j++){
                if (board[i][j]==' ') continue;
                
                if(board[i][j] == board[i][j+1] &&
                   board[i][j] == board[i+1][j] &&
                   board[i][j] == board[i+1][j+1])  
                {
                    posSet.insert(make_pair(i, j));
                    posSet.insert(make_pair(i, j+1));
                    posSet.insert(make_pair(i+1, j));
                    posSet.insert(make_pair(i+1, j+1));
                }
            }
        }

        if(posSet.size()==0) break;
        
        answer += posSet.size();
        
        for (const auto& pos : posSet)
            board[pos.first][pos.second] = ' ';

        for(int j=0; j<n; j++){
            for (int i=m-1; i>0; i--){
                if (board[i][j]==' ')
                {
                    int idx = i-1;
                    while(idx>-1 && board[idx][j]==' ')
                        idx--;
                    
                    if(idx>=0)
                        swap(board[i][j], board[idx][j]);
                }
            }
        }
    }
    
    return answer;
}